pcdscalc/diffraction.py

```diff
@@ -118,3 +121,109 @@ def d_space(material_id, hkl):
+        scattering angle in degrees
+    energy: float
+    """
+    material_id = cst.chemical_name_to_formula.get(material_id, material_id)
+    z = cst.element_z[material_id]
+    q = xraylib.MomentTransf(energy*1e-3, np.deg2rad(twotheta/2))
+    f = xraylib.FF_Rayl(z, q)
+    return f
+
+
+def structure_factor(material_id, f, hkl, z=None):
+    material_id = cst.chemical_name_to_formula.get(material_id, material_id)
+    lattice = cst.lattice_type[material_id]
+    h, k, l = hkl  # noqa: E741
+
+    if lattice == 'fcc':
+        x1 = np.exp(-1j*np.pi * (k+l))
+        x2 = np.exp(-1j*np.pi * (h+l))
+        x3 = np.exp(-1j*np.pi * (h+k))
+        F = f * (1 + x1 + x2 + x3)
+    elif lattice == 'bcc':
+        F = f * (1 + np.exp(-1j*np.pi * (h+k+l)))
+    elif lattice == 'cubic':
+        F = f
+    elif lattice == 'diamond':
+        x1 = np.exp(-1j*np.pi * (k+l))
+        x2 = np.exp(-1j*np.pi * (h+l))
+        x3 = np.exp(-1j*np.pi * (h+k))
+        x4 = np.exp(-1j*2*np.pi * (h/4.0 + k/4.0 + l/4.0))
+        F = f * (1 + x1 + x2 + x3) * (1 + x4)
+    elif lattice == 'rhomb':
+        z = cst.latticeParamRhomb[material_id]
+        F = f * (1 + np.exp(2*1j*np.pi * (h+k+l) * z))
+    elif lattice == 'tetr':
+        F = f
+    elif lattice == 'hcp':
+        F = f * (1 + np.exp(2*1j*np.pi * (h/3.0 + 2*k/3.0 + l/2.0)))
+    return F
+
+
+def unit_cell_volume(material_id):
+    """
+    Unit cell volume
+    """
+    material_id = cst.chemical_name_to_formula.get(material_id, material_id)
+    lp = cst.lattice_parameters[material_id]
+    # a, b, c in angstroms
+    # alpha, beta, gamma in degrees
+    a, b, c, alpha, beta, gamma = lp
+    a = a / units["ang"]
+    b = b / units["ang"]
+    c = c / units["ang"]
+    cos_alpha = cosd(alpha)
+    cos_beta = cosd(beta)
+    cos_gamma = cosd(gamma)
+    r = 1 - cos_alpha**2 - cos_beta**2 - cos_gamma**2 \
+        + 2 * cos_alpha * cos_beta * cos_gamma
+    volume = a * b * c * np.sqrt(r)
+    return volume
+
+
+def darwin_width(material_id, hkl, energy, T=293):
+    """
+    Computes the Darwin width for a specified crystal reflection (degrees)
+
+    Parameters
+    ----------
+    material_id : str
+        Chemical fomula. E.g.: `Si`
+    hkl : tuple
+        Miller Indices, the reflection. E.g.: `(1,1,1)`
+    E : float
+        Photon energy in eV
+    T : float
+        Temperature in Kelvin.
+
+    Returns
+    -------
+    darwin_width : float
+        Darwin width in deg of the reflection
+    """
+    material_id = cst.chemical_name_to_formula.get(material_id, material_id)
+    lam = energy_to_wavelength(energy)
+    theta = bragg_angle(material_id, hkl, energy)
+
+    f = form_factor(material_id, 2*theta, energy)
+    F = structure_factor(material_id, f, hkl)
+    vol = unit_cell_volume(material_id)
+    dw = (2 * const['eRad'] * lam**2 * np.abs(F)) \
+        / (np.pi * vol * sind(2*theta)) / units['rad']
+    print(f"""
+    f = {f}
+    F = {F}
+    vol = {vol}
+    """)
```

Comment:
Is this the same behavior as the old python? Print and also return?

Reply:
No this was some debugging to test that all variables are correct. I need to remove this!

Reply:
Consider using `logger.debug(text)` as well, instead of just removing them! Then this will end up in the debug stream, which feeds into the hutch-python logfiles.

Reply:
iirc
```
import logging
logger = logging.getLogger(__name__)
```

Reply:
I dont think this is important here. I just had a different value than the old pyhton when testing and wanted to see which variable was bad. I don't see a real use case for this moving forward